src/nbl/asset/ICPUImage.cpp

```diff
@@ -0,0 +1,319 @@
+
+			assert(CFlattenRegionsImageFilter::execute(policy, &proxy.flatten)); // this should never fail, at this point we are already validated
+		}
+
+		/*
+			now when the output is prepared we go with single
+			region covering all texels for a given mip level & layers,
+			note if we have optimized (fully flattened) input we skip
+			flattening by hand to avoid extra allocations & copies
+		*/
+
+		const auto* const image = isFullyFlatten ? state->inImage : proxy.flatten.outImage.get();
+		const auto texelOrBlockByteSize = asset::getTexelOrBlockBytesize(parameters.format);
+		const uint8_t* inData = reinterpret_cast<const uint8_t*>(image->getBuffer()->getPointer());
+
+		std::vector<uint32_t> layers(parameters.arrayLayers);
+		std::iota(layers.begin(), layers.end(), 0);
+
+		std::vector<uint32_t> levels(parameters.mipLevels);
+		std::iota(levels.begin(), levels.end(), 0);
+
+		/*
+			we stream-hash texels per given mip level & layer
+		*/
+
+		auto* const hashes = reinterpret_cast<CState::hash_t*>(getScratchAsBuffer(scratch.hashes.size, scratch.hashes.offset)->getPointer());
+		auto* const hashers = reinterpret_cast<blake3_hasher*>(getScratchAsBuffer(scratch.hashers.size, scratch.hashers.offset)->getPointer());
+
+		auto executePerMipLevel = [&](const uint32_t miplevel)
+		{
+			const auto mipOffset = (miplevel * parameters.arrayLayers);
+
+			auto executePerLayer = [&](const uint32_t layer)
+			{
+				const auto pOffset = mipOffset + layer;
+				auto* const hasher = hashers + pOffset;
+				auto* const hash = hashes + pOffset;
+
+				blake3_hasher_init(hasher);
+
+				IImage::SSubresourceLayers subresource = { .aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0u), .mipLevel = miplevel, .baseArrayLayer = layer, .layerCount = 1u }; // stick to given mip level and single layer
+				CMatchedSizeInOutImageFilterCommon::state_type::TexelRange range = { .offset = {}, .extent = { parameters.extent.width, parameters.extent.height, parameters.extent.depth } }; // cover all texels within layer range, take 0th mip level size to not clip anything at all
+				CBasicImageFilterCommon::clip_region_functor_t clipFunctor(subresource, range, parameters.format);
+
+				auto executePerTexelOrBlock = [&](uint32_t readBlockArrayOffset, core::vectorSIMDu32 readBlockPos) -> void
+				{
+					blake3_hasher_update(hasher, inData + readBlockArrayOffset, texelOrBlockByteSize);
+				};
+
+				const auto regions = image->getRegions(miplevel);
+				const bool performNullHash = regions.empty();
+
+				if (performNullHash)
+				{
+					const auto mipExtentInBlocks = info.convertTexelsToBlocks(image->getMipSize(miplevel));
+					const auto zeroLength = info.getBlockByteSize() * mipExtentInBlocks.x;
+					auto zeroArray = std::make_unique<uint8_t[]>(zeroLength);
+					for (auto z = 0; z < mipExtentInBlocks.z; z++)
+						for (auto y = 0; y < mipExtentInBlocks.y; y++)
+							blake3_hasher_update(hasher, zeroArray.get(), zeroLength);
```

Comment:
actually you have my blessing to give this a special treatment at the start, you can do a hash of a zero-length array if there are no regions (it will be much faster)

and assert() in the filter that there are SOME regions